sstables/storage.cc

```diff
@@ -114,6 +117,38 @@ future<> filesystem_storage::rename_new_file(const sstable& sst, sstring from_na
     });
 }
 
+future<> filesystem_storage::sync_cached_directory(std::string_view name) const {
+    std::string key { name };
+    auto iter { _dircache.find(key) };
+    file dir_f;
+
+    if (iter == _dircache.end()) {
+        dir_f = co_await open_directory(name);
+
+        try {
+            auto res { _dircache.emplace(std::move(key), std::move(dir_f)) };
+
+            assert(res.second);
+            iter = res.first;
+        } catch (...) {
+            sstlog.warn("Failed to insert {} into cache: {}", key, std::current_exception());
+        }
+    }
+
+    if (iter == _dircache.end()) {
+        co_await dir_f.flush();
```

Comment:
That was precisely what I did at first -- only to find out that `co_await` is forbidden in exception handler blocks.

https://en.cppreference.com/w/cpp/language/coroutines

    A co_await expression can only appear in a potentially-evaluated
    expression within a regular function body (including the function
    body of a lambda expression), and cannot appear

        in a handler,

The code, as-submitted, is effectively a workaround for that language restriction.

Reply:
`if (dir_f)` should work here too, instead of checking the cached_dir iterator.
I think it's clearer since the code in this block uses dir_f while the code below it uses cached_dir.
In any case, it's worth a shott comment explaining that the block handles the error above.

Another way we commonly use in coroutines is defining `std::exception_ptr ex` before the try/catch block and assign it with `std::current_exception()` inside the catch clause.
And then check for `if (ex)` outside the catch block.
This clearly shows that the code is doing error handling.

Reply:
> `if (dir_f)` should work here too, instead of checking the cached_dir iterator. I think it's clearer since the code in this block uses dir_f while the code below it uses cached_dir. In any case, it's worth a shott comment explaining that the block handles the error above.

Evaluating just `dir_f` here is insufficient. `dir_f` will eval to `true` here if the cache lookup failed (because then we open the directory, and assign `dir_f` -- and if that throws, then we don't reach this `if`.) The condition for running the `flush`+`close` sequence is that the directory not be cached originally, **and** that we also fail to insert it into the cache. This is stricter than just `dir_f` being set.

The iterator not pointing inside the container is a precise condition; it will only hold if the initial lookup failed and then the insertion failed too.

> Another way we commonly use in coroutines is defining `std::exception_ptr ex` before the try/catch block and assign it with `std::current_exception()` inside the catch clause. And then check for `if (ex)` outside the catch block. This clearly shows that the code is doing error handling.

Thanks, I'll implement this (this lets me / requires me to move some stuff under the scope of "directory not found in cache). With the comment of course.

Reply:
... not sure what the proper etiquette is -- some reviewers prefer the patch author to resolve threads whenever the author thinks they have addressed the concern; other reviewers prefer to verify the code changes and/or answers and then resolve the threads themselves. For now I'm going to resolve the threads that I think to have addressed, because the github UI is getting cluttered and I find it difficult to navigate it (yearning for good ol' mailing list based review sessions! ;) )

Reply:
> ... not sure what the proper etiquette is -- some reviewers prefer the patch author to resolve threads whenever the author thinks they have addressed the concern; other reviewers prefer to verify the code changes and/or answers and then resolve the threads themselves. For now I'm going to resolve the threads that I think to have addressed, because the github UI is getting cluttered and I find it difficult to navigate it (yearning for good ol' mailing list based review sessions! ;) )

Yeah, it was discussed elsewhere.  There's no one answer for that in scylla.
It makes sense for the reviewer to resolve their comments when they were satisfied.
As a reviewee, when I comment on a thread that I addressed, I don't resolve it immediately, so to let the reviewer a chance to read the comment and decide if they are satisfied and resolve it.
Then, when I generate another version, I resolve the old threads that were already addressed in previous version so they won't clutter the issue, if the reviewer hadn't done so yet.

Reply:
> As a reviewee, when I comment on a thread that I addressed, I don't resolve it immediately, so to let the reviewer a chance to read the comment and decide if they are satisfied and resolve it. Then, when I generate another version, I resolve the old threads that were already addressed in previous version so they won't clutter the issue, if the reviewer hadn't done so yet.

Let's say you submit `v1`, and the reviewer asks for a change. You implement the change (in your local tree), and state so under the reviewer's request. You also post `v2`, with a summary of changes in a new comment as well.

The reviewer now checks `v2`, and asks for some new (*independent*) updates, in a `v3`. They do not comment on the *previous* issue they pointed out (under `v1`), while reviewing `v2`. So you prepare `v3` (for addressing the new comments), and while doing that, you close the `v1` discussion as resolved (because you had commented on it, had showed the updated code in `v2`, and the reviewer didn't disagree). Is this what you mean?

This sounds very challenging; it requires me to track comments for *two* versions in the past.

Reply:
Yeah, I admit it is a bit challenging, other review boards I've seen had better management of review comments where they needed to be explicitly resolved in order to merge a change.

> Let's say you submit v1, and the reviewer asks for a change. You implement the change (in your local tree), and state so under the reviewer's request. You also post v2, with a summary of changes in a new comment as well.

I document the summary of changes at the bottom of the GH issue, and also comment on each review comment I addresses with a short status like "done in v2", or "no change in v2 since this is still controversial".
This way it is easier to go back and resolve those threads, if there were no further comments after your last update.

Reply:
Ah, adding separate notes with version references to the open discussion threads is helpful. Thanks!